## 이메일 에러를 통한 에러처리 예제

- 발생 할 수 있는 오류들을 나열 해보면 아래와 같다.
```swift 
import UIKit

enum ValidationError: Error {
    case noEmptyValueAllowed
    case invalidEmail
}
```

- 이메일 검증을 하는 로직을 만들고, 에러가 발생할 가능성이 존재하기 때문에 throws를 선언한다.

```swift 
func validateEmail(_ email: String) throws {
    
    guard !email.isEmpty else {
        throw ValidationError.noEmptyValueAllowed
    }
    
    let pattern = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
    
    let emailPredicate = NSPredicate(format: "SELF MATCHES %@", pattern)
    let isValid = emailPredicate.evaluate(with: email)
    if !isValid {
        throw ValidationError.invalidEmail
    } 
}
```

- 에러가 발생할 수 있는 가능성이 있는 함수를 실행 시키기 위해 do-catch문을 사용한다.

```swift 
do {
    try validateEmail("")
    print("Email is valid")
} catch {
    print(error)
}
```

- 구조체를 통해서 이메일 검증을 진행해보자
```swift 
struct Email {
    let text: String
    
    init(_ text: String) throws {
        
        guard !text.isEmpty else {
            throw ValidationError.noEmptyValueAllowed
        }
        
        let pattern = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let emailPredicate = NSPredicate(format: "SELF MATCHES %@", pattern)
        
        let isValid = emailPredicate.evaluate(with: text)
        if isValid {
            self.text = text
        } else {
            throw ValidationError.invalidEmail
        }
    }
}

do {
    let email = try Email("johndoe@gmail.com")
    print(email)
} catch {
    print(error)
}
```
